src/commands/extract.rs

```diff
@@ -160,53 +167,87 @@ impl Extract {
-                }
+impl Resources {
+    pub fn execute(&self) -> Result<()> {
+        if self.paths.is_empty() {
+            bail!("Input path does not exist")
+        }
 
-                load_trust_settings(trust)?;
+        if !self.output.exists() {
+            fs::create_dir_all(&self.output)?;
+        } else if !self.output.is_dir() {
+            bail!("Output path must be a folder");
+        } else if !self.force {
+            // TODO: if self.force is specified, shuld we clear the folder?
+            bail!("Output path already exists use `--force` to overwrite and clear children");
+        }
 
-                let mut errs = Vec::new();
-                for path in paths {
-                    if path.is_dir() {
-                        bail!("Input path cannot be a folder when extracting resources");
-                    }
+        load_trust_settings(&self.trust)?;
 
-                    if let Err(err) = ManifestStore::from_file_with_resources(path, output) {
-                        error!(
-                            "Failed to extract resources from asset at path `{}`, {}",
-                            path.display(),
-                            err.to_string()
-                        );
-                        errs.push(err);
-                    }
-                }
+        let mut errs = Vec::new();
+        for path in &self.paths {
+            if path.is_dir() {
+                bail!("Input path cannot be a folder when extracting resources");
+            }
 
-                if !errs.is_empty() {
-                    bail!(
-                        "Failed to extract resources from {}/{} assets",
-                        errs.len(),
-                        paths.len()
-                    );
+            if let Err(err) = self.extract_resources(path) {
+                error!(
+                    "Failed to extract resources from asset at path `{}`, {}",
+                    path.display(),
+                    err.to_string()
+                );
+                errs.push(err);
+            }
+        }
+
+        if !errs.is_empty() {
+            bail!(
+                "Failed to extract resources from {}/{} assets",
+                errs.len(),
+                self.paths.len()
+            );
+        }
+
+        Ok(())
+    }
+
+    fn extract_resources(&self, path: &Path) -> Result<()> {
+        let reader = Reader::from_file(path)?;
+        for manifest in reader.iter_manifests() {
+            let manifest_path = self.output.join(
+                manifest
+                    .label()
+                    .context("Failed to get maniest label")?
+                    .replace(':', "_"),
+            );
+            for resource_ref in manifest.iter_resources() {
+                if !self.unknown && resource_ref.format == "application/octet-stream" {
+                    continue;
                 }
+
+                let uri = self.normalize_uri(&resource_ref.identifier);
```

Comment:
This doesn't do enough. URIs can have colons in them which will do bad things in OS file paths. They should be converted to underscores.

Reply:
Take a look at what the code in ResourceStore.rs does.

Reply:
I added most of the code from ResourceStore. One thing is file extension inference found [here](https://github.com/contentauth/c2pa-rs/blob/2aeafd3888a6b96d00543d29a58c9783f6785f31/sdk/src/utils/mime.rs#L62), but I'm not sure if it's reasonable to copy all of that code.